app/models/firmware.rb

```diff
@@ -3,6 +3,6 @@ class Firmware < ApplicationRecord
 
   acts_as_miq_taggable
 
-  belongs_to :hardware, :polymorphic => true
+  belongs_to :hardware, :polymorphic => true, :foreign_key => :resource_id, :foreign_type => :resource_type
```

Comment:
It's a bit misleading that `Firmware#hardware` could return something that is not a Hardware.